db/blip_handler.go

```diff
@@ -792,65 +823,107 @@ func (bh *blipHandler) handleGetAttachment(rq *blip.Message) error {
-				nonce, proof := GenerateProofOfAttachment(knownData)
-				outrq := blip.NewRequest()
-				outrq.Properties = map[string]string{BlipProfile: MessageProveAttachment, ProveAttachmentDigest: digest}
-				outrq.SetBody(nonce)
-				if !bh.sendBLIPMessage(sender, outrq) {
-					return nil, ErrClosedBLIPSender
-				}
-				if body, err := outrq.Response().Body(); err != nil {
-					base.WarnfCtx(bh.loggingCtx, "Error returned for proveAttachment message for doc %s (digest %s).  Error: %v", base.UD(docID), digest, err)
-					return nil, err
-				} else if string(body) != proof {
-					base.WarnfCtx(bh.loggingCtx, "Incorrect proof for attachment %s : I sent nonce %x, expected proof %q, got %q", digest, base.MD(nonce), base.MD(proof), base.MD(string(body)))
-					return nil, base.HTTPErrorf(http.StatusForbidden, "Incorrect proof for attachment %s", digest)
-				} else {
-					base.InfofCtx(bh.loggingCtx, base.KeySync, "proveAttachment successful for doc %s (digest %s)", base.UD(docID), digest)
-				}
-				return nil, nil
-			} else {
-				// If I don't have the attachment, I will request it from the client:
-				base.DebugfCtx(bh.loggingCtx, base.KeySync, "    Asking for attachment %q for doc %s (digest %s)", base.UD(name), base.UD(docID), digest)
-				outrq := blip.NewRequest()
-				outrq.Properties = map[string]string{BlipProfile: MessageGetAttachment, GetAttachmentDigest: digest}
-				if isCompressible(name, meta) {
-					outrq.Properties[BlipCompress] = "true"
-				}
-				if !bh.sendBLIPMessage(sender, outrq) {
-					return nil, ErrClosedBLIPSender
-				}
-				attBody, err := outrq.Response().Body()
-				if err != nil {
-					return nil, err
-				}
 
-				lNum, metaLengthOK := meta["length"].(json.Number)
-				metaLength, err := lNum.Int64()
-				if err != nil {
-					return nil, err
-				}
+			haveAttachment := knownData != nil
 
-				// Verify that the attachment we received matches the metadata stored in the document
-				if !metaLengthOK || len(attBody) != int(metaLength) || Sha1DigestKey(attBody) != digest {
-					return nil, base.HTTPErrorf(http.StatusBadRequest, "Incorrect data sent for attachment with digest: %s", digest)
-				}
+			if !haveAttachment {
+				return bh.sendGetAttachment(sender, docID, name, digest, meta)
+			}
 
-				bh.replicationStats.GetAttachment.Add(1)
-				bh.replicationStats.GetAttachmentBytes.Add(metaLength)
-				return attBody, nil
+			proveAttErr := bh.sendProveAttachment(sender, docID, name, digest, knownData)
+			if proveAttErr == NoBLIPHandlerError {
+				// fall back to getAttachment (we're probably replicating with a pre-Hydrogen SG, which doesn't support proveAttachment)
+				_, getAttErr := bh.sendGetAttachment(sender, docID, name, digest, meta)
+				if getAttErr != nil {
+					return nil, getAttErr
+				}
+			} else if proveAttErr != nil {
+				return nil, proveAttErr
 			}
+
+			return nil, nil
```

Comment:
Maybe move this up as an early return when proveAttErr== nil?  It took me a minute to work out that this was the successful prove return.